src/addons/conditionwatchers/addonconditionwatchertimeend.cpp

```diff
@@ -4,23 +4,24 @@
 
 #include "addonconditionwatchertimeend.h"
 #include "leakdetector.h"
+#include "mfbt/checkedint.h"
 #include "settingsholder.h"
 
 #include <QDateTime>
 
 AddonConditionWatcherTimeEnd::AddonConditionWatcherTimeEnd(QObject* parent,
                                                            qint64 time)
-    : AddonConditionWatcher(parent) {
+    : AddonConditionWatcher(parent), m_time(time) {
   MVPN_COUNT_CTOR(AddonConditionWatcherTimeEnd);
 
-  qint64 currentTime = QDateTime::currentSecsSinceEpoch();
-  if (time > currentTime) {
-    m_timer.setSingleShot(true);
-    m_timer.start((time - currentTime) * 1000);
+  m_timer.setSingleShot(true);
+  connect(&m_timer, &QTimer::timeout, this, [this]() {
+    if (maybeStartTimer()) {
+      emit conditionChanged(false);
+    }
+  });
 
-    connect(&m_timer, &QTimer::timeout, this,
-            [this]() { emit conditionChanged(false); });
-  }
+  maybeStartTimer();
```

Comment:
Don't we need to emit a `conditionChanged` if this returns `true`? What is the condition is already met when the condition watcher is started?

Reply:
no signals in the CTOR. Nobody will receive them.

Reply:
Ah, ok. But still what happens if the condition is already met when the condition watcher is started?

Reply:
we call: `conditionApplied()` after the CTOR.